src/plugins/expressions/common/expression_functions/specs/moving_average.ts

```diff
@@ -0,0 +1,161 @@
+ * * [window] of values divided by [window] excluding the current bucket.
+ * If either of window edges moves outside the borders of data series, the window shrinks to include available values only.
+ *
+ * Edge cases:
+ * * Will return the input table if `inputColumnId` does not exist
+ * * Will throw an error if `outputColumnId` exists already in provided data table
+ * * If null or undefined value is encountered, skip the current row and do not change the window
+ * * For all values besides `null` and `undefined`, the value will be cast to a number before it's used in the
+ *   calculation of the current series even if this results in `NaN` (like in case of objects).
+ * * To determine separate series defined by the `by` columns, the values of these columns will be cast to strings
+ *   before comparison. If the values are objects, the return value of their `toString` method will be used for comparison.
+ */
+export const movingAverage: ExpressionFunctionMovingAverage = {
+  name: 'moving_average',
+  type: 'datatable',
+
+  inputTypes: ['datatable'],
+
+  help: i18n.translate('expressions.functions.movingAverage.help', {
+    defaultMessage: 'Calculates the moving average of a column in a data table',
+  }),
+
+  args: {
+    by: {
+      help: i18n.translate('expressions.functions.movingAverage.args.byHelpText', {
+        defaultMessage: 'Column to split the moving average calculation by',
+      }),
+      multi: true,
+      types: ['string'],
+      required: false,
+    },
+    inputColumnId: {
+      help: i18n.translate('expressions.functions.movingAverage.args.inputColumnIdHelpText', {
+        defaultMessage: 'Column to calculate the moving average of',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnId: {
+      help: i18n.translate('expressions.functions.movingAverage.args.outputColumnIdHelpText', {
+        defaultMessage: 'Column to store the resulting moving average in',
+      }),
+      types: ['string'],
+      required: true,
+    },
+    outputColumnName: {
+      help: i18n.translate('expressions.functions.movingAverage.args.outputColumnNameHelpText', {
+        defaultMessage: 'Name of the column to store the resulting moving average in',
+      }),
+      types: ['string'],
+      required: false,
+    },
+    window: {
+      help: i18n.translate('expressions.functions.movingAverage.args.windowHelpText', {
+        defaultMessage: 'The size of window to "slide" across the histogram.',
+      }),
+      types: ['number'],
+      required: false,
+      default: 5,
+    },
+  },
+
+  fn(input, { by, inputColumnId, outputColumnId, outputColumnName, window = 5 }) {
+    const resultColumns = buildResultColumns(
+      input,
+      outputColumnId,
+      inputColumnId,
+      outputColumnName
+    );
+
+    if (!resultColumns) {
+      return input;
+    }
+
+    const lastNValuesByBucket: Partial<Record<string, Array<number | undefined>>> = {};
```

Comment:
If I'm reading this correctly, no `undefined` values will end up in `lastNValuesByBucket`, right? If that's the case the algorithm can be simplified by removing `sanitizedLastNValues` and just using `lastNValues` directly.